plutus-core/plutus-core/src/PlutusCore/Builtin/KnownType.hs

```diff
@@ -100,42 +100,46 @@ the term is a constant, extract the type tag and check it for equality with the
 Things work the same way for a fully monomorphized polymorphic type, i.e. @(Integer, Bool@) is not
 any different from just 'Integer' unlifting-wise.
 
-However there's no sensible way of unlifting a value of, say, @[a]@ where @a@ in not a built-in
-type. So let's say we instantiated @a@ to an @Opaque val rep@ like we do for polymorphic functions
-that don't deal with polymorphic built-in types (e.g. @id@, @ifThenElse@ etc). That would mean we'd
-need to write a function from a @[a]@ for some arbitrary built-in @a@ to @[Opaque val a]@. Which
-is really easy to do: it's just @map makeKnown@. But the problem is, unlifting is supposed to be
-cheap and that @map@ is O(n), so for example 'MkCons' would become an O(n) operation making
+When it comes to polymorphic types, unlifting a @val@ into an @Opaque val a@ is trivial - it's just
+@makeKnown@. However, there's no sensible way of unlifting a @val@ into, say, @[Opaque val a]@
+(which would be needed if we use the same approach in defining @NullList@, @MkCons@ etc. as we do
+defining @IfThenElse@ and @ChooseUnit@), because we'd need to first unlift the @val@ into a
+constant, then check that it is indeed a list, and finally @map makeKnown@ over the list. But the problem is,
+unlifting is supposed to be cheap, while mapping over a list is O(n). This isn't a problem for
+@NullList@ because it only needs to examine whether the list is empty, but it is a problem for @MkCons@
```

Comment:
No, that's the whole point! `NullList` is as problematic as `MkCons` in that regard. It doesn't matter if a built-in function consumes a list or produces one, if that involves any `[Opaque val a]`, we have to use a `map` and this is where the linearity comes from.

Reply:
How so, if we never need the elements of the list? Like `null . map f` is O(1) just like `null` is?